examples/bq-email-exports/main.py

```diff
@@ -49,87 +51,54 @@ def credentials():
 
-    # Set variables
-    timestr = time.strftime("%Y%m%d%I%M%S")
-    project = "report-scheduling"
-    dataset_id = "bq_exports"
-    file_name = "daily_export_" + timestr
-    csv_name = file_name + ".csv"
-    table_id = "report-scheduling.bq_exports." + file_name
-    bucket_name = "bq_email_exports"
-    from_email = "sender@example.com"
-    to_emails = "recipient@example.com"
-
-    # Create a BQ table
-    schema = [
-        bigquery.SchemaField("url", "STRING", mode="REQUIRED"),
-        bigquery.SchemaField("view_count", "INTEGER", mode="REQUIRED"),
-    ]
-
-    # Make an API request to create table
-    table = bq_client.create_table(bigquery.Table(table_id, schema=schema))
-    print("Created table {}.{}.{}".format(table.project, table.dataset_id,
-                                          table.table_id))
-
-    # Run query on that table
-    job_config = bigquery.QueryJobConfig(destination=table_id)
+    # Get configurations from Cloud Scheduler payload
+    config = json.loads(base64.b64decode(event['data']).decode('utf-8'))
 
-    # Define the SQL query
-    sql = """
-        SELECT
-        CONCAT(
-            'https://stackoverflow.com/questions/',
-            CAST(id as STRING)) as url,
-        view_count
-        FROM `bigquery-public-data.stackoverflow.posts_questions`
-        WHERE tags like '%google-bigquery%'
-        ORDER BY view_count DESC
-        LIMIT 10
-    """
+    # Append timestamp to table name and set table_id 
+    table_name = config["table_name"] + time.strftime("%Y%m%d%I%M%S")
+    table_id = f"{config['project_id']}.{config['dataset_id']}.{table_name}"
+
+    # Get query 
+    job_config = bigquery.QueryJobConfig(destination=table_id)
+    query_bucket = storage_client.bucket(config["query_bucket"])
+    query_file = query_bucket.blob(config["query_file_name"])
+    query = query_file.download_as_string()
+    sql = query.decode("utf-8") 
 
     # Start the query, passing in the extra configuration
     query_job = bq_client.query(sql, job_config=job_config)
 
     # Wait for the job to complete
     query_job.result()
-    print("Query results loaded to the table {}".format(table_id))
-
-    # Export table data as CSV to GCS
-    destination_uri = "gs://{}/{}".format(bucket_name, csv_name)
-    dataset_ref = bq_client.dataset(dataset_id, project=project)
-    table_ref = dataset_ref.table(file_name)
+    print(f"Query results loaded to the table {table_id}")
 
+    # Export table data as JSON file to GCS
+    destination_uri = f"gs://{config['bucket_name']}/{table_name}.json"
+    dataset_ref = bigquery.DatasetReference(config["project_id"], config["dataset_id"])
+    table_ref = dataset_ref.table(table_name)
     extract_job = bq_client.extract_table(
         table_ref,
-        destination_uri,
-        # Location must match that of the source table
-        location="US",
+        destination_uri
     )
 
     # Waits for job to complete
     extract_job.result()
```

Comment:
Set a timeout of ~ 9 mins 50 seconds.
Reasoning:  
- if your cloud function execution is about to time out you want a chance to log to tell the user you've aborted query with job id because you wouldn't have sufficient time to export it. 
- if we can't export it there's no value in materializing the temp table.
- we should free up the slots in case of a run away query the might run for hours.

Reply:
Definitely agree with this. Based on this [doc](https://cloud.google.com/functions/docs/concepts/exec#timeout), it looks like the max timeout for a Cloud Function is 9 minutes. I was thinking I could split the 9 minutes evenly amongst the query job (Line 79) and extract job (~4 minute timeout for each). Thoughts?

Reply:
I think the appropriate timeouts depends on the query and data volume but in many cases I think the query should get the lion's share of the allotted time but it's tough to say.

## Taking a step back
### Think about the end user
It's reasonable to think that a user wants this emailed daily because the query takes a long time to run (e.g. a daily report that takes an hour to crunch).

I'm not sure how valuable a solution that only supports 4 min queries is as this is.

We shouldn't expose the user (who cares about BigQuery and GCS) to a limitation of cloud functions (and implementation detail we chose in our design).

### Think about the failure mode
We're running into a timeout issue in cloud functions so we should think about where our cloud function is spending most of its time. 
It spends most of it's time blocking on the success of the query / export jobs.
Is there value in paying for CF execution time when the CF is just waiting?
Can we trim the fat of this wait time? Yes we can see Option 2 below.

## Paths Forward
### Option 1: User Contempt (Minor refactor)
If you want a sad user experience add `queryTimeout` and `exportTimeout` to your payload and let the user worry about this, documenting that these should together be less than the Cloud Functions Timeout. This would lead to a user experience where they have to tune these timeouts and if they can't fit w/in the CF timeout window they're outta luck.

### Option 2: Event driven micro-services architecture (Major refactor)
It would be a far superior solution if this was separated into smaller fast executing cloud functions:
1) triggered by scheduler* and submits query job
2) pubsub subscribes to audit log sink with filter for query job completion (this will likely require naming scheduled query jobs with a special prefix) and runs export job
3) pubsub subscribes to audit log sink with filter for export job completion (this will likely require naming export jobs with a special prefix) and sends email / cleans up temp table.

This way your cloud functions are all small and will exit well within a few seconds but the individual jobs can take as long as they need to (within the bigquery timeouts).

Note, that waiting for the audit log to propogate through the sink would add some latency between each step but this is likely tolerable in light of the benefits.
CFs 2 and 3 would have to log a warning if their upstream job failed.


*CF (1) and Cloud Scheduler are entirely replaceable with Bigquery scheduled queries which [support a pubsub notification mechanism natively](https://googleapis.dev/python/bigquerydatatransfer/latest/gapic/v1/types.html#google.cloud.bigquery_datatransfer_v1.types.TransferConfig.notification_pubsub_topic). This would have the benefit of eliminating the dependency on cloud scheduler (which notably cannot run inside VPC-SC perimeter and depends on the AppEngine stack).
Unfortunately, this field is not well documented and not supported by the terraform resource, please +1 [this issue](https://github.com/terraform-providers/terraform-provider-google/issues/6412) or better yet patch it with a PR :).

Reply:
I agree that Option 2 is the superior solution and would eliminate the blocker of waiting on BigQuery jobs. Definitely something that @ishitashah24 and I will look into and work on implementing, given that it will take some time.

At this point, though, it may be easier to execute this as an Airflow DAG since we would be using Functions and Pub/Sub to manually run a pipeline triggered by specific logs. I'd have to look into the tradeoffs more but it could be a more logical solution.

Reply:
Airflow DAG might be easier to implement due to lots of things being done for you.
However, condiser Customer / Business interest:
Option 2 will cost less than a penny per month (sans BQ cost) because it scales to zero.
While a Composer instance will cost ~$400 / month (sans BQ cost)

Also Composer instance is a higher maintenance / operational burden than Cloud Functions.

I think it makes for a more interesting example to have scale to zero / fully managed / truly  serverless solution. These sorts of pipelines are really cutting edge and customers want to know how to do them. @kardiff18 has a cool [example](https://github.com/GoogleCloudPlatform/professional-services/tree/master/examples/ml-audio-content-profiling) in this repo of using ML APIs from CFs that uses 4 cloud functions and GCS Object change notifications to achieve a pretty neat ETL pipeline. This is really neat and probably has a similar cost savings compared to a dataflow streaming pipeline. I'd be keen to have more truly serverless data pipelines in this repo.

Reply:
Good points! Okay, we can refractor our code to work with the long running BQ queries.

Reply:
Also connected w/ tim offline. You should do you export on the bigquery anonymous table (that gets cleaned up for you and manages minimum permissions in anonymous dataset) rather than creating your own staging / temp table (which just gets exported then deleted)

Reply:
One downside of [Option #2](https://github.com/GoogleCloudPlatform/professional-services/pull/474/files/ec0872fc34cda351f21b9fc06b4ffad3152e4b94#r426979469): Cloud scheduler can send a payload with personalized query configs to Pub/Sub topic 1, but since Pub/Sub topics 2 and 3 are subscribed to the audit log sink, they will not have access to this payload. This means the user would have to customize the export job and email configs within the functions' code itself before deploying. 

Specifically, these payload configs would have to be customized within the code for function 2 (export job):
```
"bucket_name": "${bucket}",
"extract_config": {
    "compression": "NONE",
    "destination_fmt": "NEWLINE_DELIMITED_JSON",
    "use_avro": false,
    "field_delimeter": ","
  }
```
And these for function 3 (send email):
```
  "signed_url_expiration_hrs": 24,
  "from_email": "sender@example.com",
  "to_email": "recipient@example.com",
  "email_subject": "Daily BQ export"
}
```

Not a huge burden compared to the benefits, but not as clean as managing all configurations in one payload. Open to any suggestions, if any.